### ISSUE TITLE
Feature suggestion: auto correct root bone orientation?

### ISSUE THREAD

**Issue by sacrish** [xavier150/Blender-For-UnrealEngine-Addons#205]
If I manually include a root bone in my rig hierarchy, it will get a non-zero rotation value when imported to UE. This could cause problem with some retargeted animations. Setting its rotation to 0 in UE can't fix the problem because it will make the character lie on the floor. 

This seems to be the default behavior of Blender's stock fbx exporter as well.

A workaround I found is to remove the root bone from hierarchy and let the plugin automatically add a root bone on export. This way the root bone can maintain a (0,0,0) orientation in UE. But the drawback is that I would not be able to do root animations in Blender.

So I wonder whether it's possible to make sure the manually added root bone doesn't get an unapplied rotation value, or it's just hard limitation of the program?

**Comment by sacrish** [comment]
Managed to work around the problem by placing the root bone in a specific way (X pointing down and Z pointing forward) so that it matches UE's default orientation after the axis conversion:
![Screenshot 2024-12-08 150500](https://github.com/user-attachments/assets/ed38fe03-a567-43f0-95ca-5014c1bfd0bc)

I wonder if this procedure could be done by the plugin? But I think it's fine if users need to do this manually.

**Comment by xavier150** [comment]
Hey @sacrish 
You need a single root bone in any case to have a correct export or Unreal will automatically add it.
I recomand you to keep the rotation and location at Zero and Scale at 1 to avoid issues.

Unreal should keep the correct orientation in unreal. If not it's probably you need to re import the animation too.

If you need to use the Unreal bones orientation you can use the custom IO exporter here.
![image](https://github.com/user-attachments/assets/2103fdf6-599b-4c68-a1f1-2fbdf42372a3)

You can found more details here:
[UE Bone Structure and UE Mannequin](https://github.com/xavier150/Blender-For-UnrealEngine-Addons/wiki/UE-Bone-Structure-and-UE-Mannequin)
[Skeleton & Root bone](https://github.com/xavier150/Blender-For-UnrealEngine-Addons/wiki/Skeleton-&-Root-bone)

**Comment by sacrish** [comment]
> Hey @sacrish You need a single root bone in any case to have a correct export or Unreal will automatically add it. I recomand you to keep the rotation and location at Zero and Scale at 1 to avoid issues.
> 
> Unreal should keep the correct orientation in unreal. If not it's probably you need to re import the animation too.
> 
> If you need to use the Unreal bones orientation you can use the custom IO exporter here. ![image](https://private-user-images.githubusercontent.com/7216958/393594483-2103fdf6-599b-4c68-a1f1-2fbdf42372a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2NzYyMjEsIm5iZiI6MTczMzY3NTkyMSwicGF0aCI6Ii83MjE2OTU4LzM5MzU5NDQ4My0yMTAzZmRmNi01OTliLTRjNjgtYTFmMS0yZmJkZjQyMzcyYTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDhUMTYzODQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjM3MzE4ZjYwZWI5ODc2NDRkN2UyYmE4NzJkMTdkZDZkMGNmZDE3OTlkMDk0MjAxMTJmMDhhMDMwZDNkMmI2YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.gIZtV-DEj2Pz0HUH5PMqZg7n7UimO2-F-EmnndRut6Y)

Yes, I am using a single root bone hierarchy, with no unapplied transform on the Armature, and having the UE standard export selected.

My rig hierarchy is similar to UE's mannequin, like this:
![Screenshot 2024-12-09 004501](https://github.com/user-attachments/assets/4af0d52f-7322-4260-af2b-183888fa314b)
(the other CTRL_* bones are all non-deform ones so they won't be exported)

I think the thing that matters is how the root bone is oriented. For example if the root bone is placed in a conventional "Blender fashion" pointing upward like this:
![Screenshot 2024-12-09 005723](https://github.com/user-attachments/assets/7aec46ea-18ec-44b2-9643-7f87c04f90aa)

Then when imported into UE the root bone will get a default rotation:
![Screenshot 2024-12-09 010022](https://github.com/user-attachments/assets/ca50703c-5757-4689-becc-4e86f7b07dcc)

And if in Blender the root bone is placed in a specific way (so that after export axis conversion it becomes X-left Y-forward and Z-up):
![Screenshot 2024-12-09 005735](https://github.com/user-attachments/assets/d6ea4308-b9c6-41cb-9eb9-52456bcbac79)

Then when imported into UE it gets a clean transform (well almost):
![Screenshot 2024-12-09 010409](https://github.com/user-attachments/assets/053a518a-95e1-4173-9d75-64ce51736ff4)

I've attached a Blend file with both rigs so you can test them out if you're interested:
[ue_rig_test.zip](https://github.com/user-attachments/files/18053257/ue_rig_test.zip)

**Comment by xavier150** [comment]
Sorry for the delay.

If I use the "Blender fashion" pointing upward with a roll at zero I get in Unreal Engine (Pitch=90.000000,Yaw=0.000000,Roll=0.000000) It also what I get with my other skeleton from Blender.

If you need get zero for all axes well I think your method is the best.

**Comment by sacrish** [comment]
No problem. The non-zero root bone orientation in UE is fine in most cases. However, for some reason that I don't understand, some retargeted animations will have a incorrect root rotation if it's not all zero by default. 

So I imagine a feature that: the user could specify a bone as their custom root bone, and on export the plugin automatically converts its orientation (and animation values) to match UE standard. Would such a feature be possible or easy to implement? I understand if it's impossible or too much of a bother and the user has to manually set up their rig correctly.

**Comment by xavier150** [comment]
Yes I understand the problem. I would have thought that Retargeting tools could fix this. Also it is much simpler if the animations are natively compatible especially if they come from the same character.

Unfortunately an option to reorient the root bone would be too tedious, and would probably cause problems with other Rigs.
The best is to modify your armature to directly place the root bone with the right axes so as to have a value of 0 in Unreal Engine.
@SAM-tak worked on the custom fbx I/O. Maybe they has an opinion about it?

About animating the root bone I guess for the Root Motion it's better to do it with the rig animation.
Here my controller will move and rotate the root bone:
![image](https://github.com/user-attachments/assets/602869ab-bcaf-49f8-b5f3-0d437817c5e9)

**Comment by SAM-tak** [comment]
#206

I thought that this slight change might solve the issue, so I tried it, but it might not actually be the solution...

The mesh root bone's base transform no longer has any rotation, but the pose bone still has a slight rotation. It's impossible to eliminate this because it's difficult to distinguish between the intended rotation values by mechanicaly.

**Comment by SAM-tak** [comment]
tried a bit, but it wasn't easy...

I thought it might work if the process around line 3528 of `fbxio/io_scene_fbx_*/export_fbx_bin.py` included the root bone as well.

**Comment by drpsyko101** [comment]
Option for zeroing out/transform root would also be helpful with the use of the default generated rig from Blender's metarig. My pipeline does need Rigify retargeting in Blender and UE so having a coherent root transform in both workflows really helps.

**Comment by xavier150** [comment]
Hello! 
I added more info here: https://github.com/xavier150/Blender-For-UnrealEngine-Addons/wiki/Skeleton-&-Root-bone
It just exemple presets as you did with the bone rotation.
